File: 402.移掉k位数字.java
/*
 * @lc app=leetcode.cn id=402 lang=java
 *
 * [402] 移掉K位数字
 */

// @lc code=start
class Solution {
    public String removeKdigits(String num, int k) {
        if(num.length()==0||num.length()<=k)
            return "0";
        char[] strStack =new char[num.length()];
        int newLength=num.length()-k;
        int top=0;
        for(int i=0;i<num.length();i++){
            char c=num.charAt(i);
            while(top>0&&strStack[top-1]>c&&k>0){
                top--;
                k--;
            }
            strStack[top++]=c;
        }
        int index=0;
        while(index<newLength&&strStack[index]=='0')
            index++;
        return index==newLength?"0":new String(strStack,index,newLength-index);
    }
}
// @lc code=end

File: 402.移掉k位数字.py
#
# @lc app=leetcode.cn id=402 lang=python3
#
# [402] 移掉K位数字
#

# @lc code=start
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        # is_active = [True for _ in range(len(num))]
        # tmp = [int(num) for _ in range(len(num))]
        # Min = int(num)
        # index = 0
        # for i in range(k):
        #     if int(tmp[:i]+tmp[-1:-i:-1]) < Min:
        #         index = i
        out=[]
        for d in num:
            while k and out and out[-1]>d:
                out.pop()
                k-=1
            out.append(d)
        return ''.join(out[:-k or None]).lstrip('0') or '0'



        # def dp(num):
        #     dp = [int('inf') for i in range(len(num))]
        #     for i in range(len(num)):
        #         dp[i] = min()
        
# @lc code=end

